### ISSUE TITLE
Port code to ROS Kinetic and Ubuntu Xenial

### ISSUE THREAD

**Issue by jack-oquin** [utexas-bwi/segbot_arm#50]
We are trying to keep our code working simultaneously on Indigo and Kinetic. So far, that has been possible.

No work has been done yet for this repository. It depends on progress in the lower-level repositories. Then Xenial needs to be installed in the arm robot, so it can dual-boot with Trusty.

xref: utexas-bwi/bwi#44, utexas-bwi/segbot#77, utexas-bwi/bwi_common#83

**Comment by jack-oquin** [comment]
Kinetic is working to some degree now for the robot base, it's possible to start building segbot_arm, and fixing whatever problems arise.

**Comment by jack-oquin** [comment]
Checked out a copy of the master branch, extending a Kinetic BWI workspace. The first problem is resolving the dependencies:
```
ERROR: the following packages/stacks could not have their rosdep keys resolved
to system dependencies:
segbot_arm_joystick: Cannot locate rosdep definition for [kinova_msgs]
segbot_arm_tasks: Cannot locate rosdep definition for [kinova_msgs]
obstacle_avoidance: Cannot locate rosdep definition for [kinova_msgs]
segbot_arm_manipulation: Cannot locate rosdep definition for [kinova_msgs]
m1n6s200_segbot_moveit_config: Cannot locate rosdep definition for [kinova_description]
moveit_utils: Cannot locate rosdep definition for [moveit_ros_benchmarks_gui]
mimic_motion: Cannot locate rosdep definition for [kinova_msgs]
segbot_arm_perception: Cannot locate rosdep definition for [ml_classifiers]
draw_demo: Cannot locate rosdep definition for [kinova_msgs]
segbot_arm_tutorials: Cannot locate rosdep definition for [kinova_msgs]
```
Many of the missing dependencies are contained in the ``kinova_ros`` repository. Adding that to the workspace, gives these errors:
```
ERROR: the following packages/stacks could not have their rosdep keys resolved
to system dependencies:
kinova_arm_moveit_demo: Cannot locate rosdep definition for [pr2_moveit_plugins]
segbot_arm_perception: Cannot locate rosdep definition for [ml_classifiers]
segbot_arm_manipulation: Cannot locate rosdep definition for [agile_grasp]
obstacle_avoidance: Cannot locate rosdep definition for [agile_grasp]
moveit_utils: Cannot locate rosdep definition for [moveit_ros_benchmarks_gui]
```
There is no Kinetic release of ``agile_grasp`` yet. There is a ``moveit_ros_benchmarks`` package, but not ``moveit_ros_benchmarks_gui``. Adding ``agile_grasp`` to the workspace yields:
```
ERROR: the following packages/stacks could not have their rosdep keys resolved
to system dependencies:
kinova_arm_moveit_demo: Cannot locate rosdep definition for [pr2_moveit_plugins]
segbot_arm_perception: Cannot locate rosdep definition for [ml_classifiers]
moveit_utils: Cannot locate rosdep definition for [moveit_ros_benchmarks_gui]
```

**Comment by jack-oquin** [comment]
Scott Niekum wrote the ``ml_classifiers`` package. Adding it to the workspace resolves that dependency.

**Comment by jack-oquin** [comment]
After some confusion, I discovered the README instructions for segbot_arm.
```
sudo apt-get install ros-kinetic-moveit-* libsndfile1-dev libfftw3-dev liblapack-dev ros-kinetic-trac-ik
```
Here's my initial .rosinstall:
```
- git:
    local-name: agile_grasp
    uri: https://github.com/jsinapov/agile_grasp.git
- git:
    local-name: kinova_ros
    uri: https://github.com/utexas-bwi/kinova-ros.git
- git:
    local-name: ml_classifiers
    uri: https://github.com/sniekum/ml_classifiers.git
- git:
    local-name: segbot_arm
    uri: https://github.com/utexas-bwi/segbot_arm.git
```
At some point these dependencies should be provided directly via ``package.xml`` dependencies and a rosinstall file for each platform.

**Comment by jack-oquin** [comment]
Building the packages described above yields several problems:
```
Errors     << j2s7s300_ikfast:make /home/joq/ros/arm/logs/j2s7s300_ikfast/build.make.001.log
In file included from /opt/ros/kinetic/include/moveit/macros/class_forward.h:40:0,
                 from /opt/ros/kinetic/include/moveit/kinematics_base/kinematics_base.h:42,
                 from /home/joq/ros/arm/src/kinova_ros/kinova_moveit/inverse_kinematics_plugins/ikfast/j2s7s300_ikfast/src/j2s7s300_robot_arm_ikfast_moveit_plugin.cpp:49:
/opt/ros/kinetic/include/moveit/macros/declare_ptr.h:52:16: error: ‘shared_ptr’ in namespace ‘std’ does not name a template type
   typedef std::shared_ptr<Type> Name##Ptr;                                                                             \
                ^
/opt/ros/kinetic/include/moveit/macros/class_forward.h:51:3: note: in expansion of macro ‘MOVEIT_DECLARE_PTR’
   MOVEIT_DECLARE_PTR(C, C);
   ^
/opt/ros/kinetic/include/moveit/kinematics_base/kinematics_base.h:139:1: note: in expansion of macro ‘MOVEIT_CLASS_FORWARD’
 MOVEIT_CLASS_FORWARD(KinematicsBase);
 ^
/opt/ros/kinetic/include/moveit/macros/declare_ptr.h:53:16: error: ‘shared_ptr’ in namespace ‘std’ does not name a template type
   typedef std::shared_ptr<const Type> Name##ConstPtr;
                ^
/opt/ros/kinetic/include/moveit/macros/class_forward.h:51:3: note: in expansion of macro ‘MOVEIT_DECLARE_PTR’
   MOVEIT_DECLARE_PTR(C, C);
   ^
/opt/ros/kinetic/include/moveit/kinematics_base/kinematics_base.h:139:1: note: in expansion of macro ‘MOVEIT_CLASS_FORWARD’
 MOVEIT_CLASS_FORWARD(KinematicsBase);
 ^
make[2]: *** [CMakeFiles/j2s7s300_robot_arm_moveit_ikfast_plugin.dir/src/j2s7s300_robot_arm_ikfast_moveit_plugin.cpp.o] Error 1
make[1]: *** [CMakeFiles/j2s7s300_robot_arm_moveit_ikfast_plugin.dir/all] Error 2
make: *** [all] Error 2
cd /home/joq/ros/arm/build/j2s7s300_ikfast; catkin build --get-env j2s7s300_ikfast | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
.......................................................................................
Failed     << j2s7s300_ikfast:make                         [ Exited with code 2 ]      
Failed    <<< j2s7s300_ikfast                              [ 6.6 seconds ]             
```
There are similar errors in ``j2n6s300_ikfast``, and ``m1n6s300_ikfast``. The ``std::shared_ptr`` reference is a C++11 feature, so CMake probably needs to be configured to use that version of the language.

```
Warnings   << agile_grasp:cmake /home/joq/ros/arm/logs/agile_grasp/build.cmake.001.log 
CMake Warning at /opt/ros/kinetic/share/cmake_modules/cmake/Modules/FindEigen.cmake:62 (message):
  The FindEigen.cmake Module in the cmake_modules package is deprecated.

  Please use the FindEigen3.cmake Module provided with Eigen.  Change
  instances of find_package(Eigen) to find_package(Eigen3).  Check the
  FindEigen3.cmake Module for the resulting CMake variable names.
```
```
Errors     << agile_grasp:make /home/joq/ros/arm/logs/agile_grasp/build.make.001.log   
In file included from /usr/include/eigen3/Eigen/Core:386:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /home/joq/ros/arm/src/agile_grasp/include/agile_grasp/finger_hand.h:35,
                 from /home/joq/ros/arm/src/agile_grasp/src/agile_grasp/finger_hand.cpp:1:
/usr/include/eigen3/Eigen/src/Core/Redux.h: In instantiation of ‘typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::sum() const [with Derived = Eigen::Array<bool, 1, -1>; typename Eigen::internal::traits<T>::Scalar = bool]’:
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/finger_hand.cpp:132:70:   required from here
/usr/include/eigen3/Eigen/src/Core/Redux.h:449:66: warning: ‘Eigen::internal::scalar_sum_op<bool>::scalar_sum_op()’ is deprecated [-Wdeprecated-declarations]
   return derived().redux(Eigen::internal::scalar_sum_op<Scalar>());
                                                                  ^
In file included from /usr/include/eigen3/Eigen/Core:337:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /home/joq/ros/arm/src/agile_grasp/include/agile_grasp/finger_hand.h:35,
                 from /home/joq/ros/arm/src/agile_grasp/src/agile_grasp/finger_hand.cpp:1:
/usr/include/eigen3/Eigen/src/Core/functors/BinaryFunctors.h:50:3: note: declared here
   scalar_sum_op() {}
   ^
In file included from /usr/include/eigen3/Eigen/Core:386:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /home/joq/ros/arm/src/agile_grasp/include/agile_grasp/rotating_hand.h:35,
                 from /home/joq/ros/arm/src/agile_grasp/src/agile_grasp/rotating_hand.cpp:1:
/usr/include/eigen3/Eigen/src/Core/Redux.h: In instantiation of ‘typename Eigen::internal::traits<T>::Scalar Eigen::DenseBase<Derived>::sum() const [with Derived = Eigen::Array<bool, 1, -1>; typename Eigen::internal::traits<T>::Scalar = bool]’:
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/rotating_hand.cpp:111:34:   required from here
/usr/include/eigen3/Eigen/src/Core/Redux.h:449:66: warning: ‘Eigen::internal::scalar_sum_op<bool>::scalar_sum_op()’ is deprecated [-Wdeprecated-declarations]
   return derived().redux(Eigen::internal::scalar_sum_op<Scalar>());
                                                                  ^
In file included from /usr/include/eigen3/Eigen/Core:337:0,
                 from /usr/include/eigen3/Eigen/Dense:1,
                 from /home/joq/ros/arm/src/agile_grasp/include/agile_grasp/rotating_hand.h:35,
                 from /home/joq/ros/arm/src/agile_grasp/src/agile_grasp/rotating_hand.cpp:1:
/usr/include/eigen3/Eigen/src/Core/functors/BinaryFunctors.h:50:3: note: declared here
   scalar_sum_op() {}
   ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp: In member function ‘std::vector<GraspHypothesis> Learning::classify(const std::vector<GraspHypothesis>&, const string&, const Matrix3Xd&, bool)’:
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:181:2: error: ‘CvSVM’ was not declared in this scope
  CvSVM svm;
  ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:185:3: error: ‘svm’ was not declared in this scope
   svm.load(svm_filename.c_str());
   ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:193:45: error: ‘svm’ was not declared in this scope
   std::cout << " # of support vectors: " << svm.get_support_vector_count() << "\n";
                                             ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp: In member function ‘void Learning::convertData(const std::vector<Learning::Instance>&, const string&, bool, bool)’:
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:298:2: error: ‘CvSVMParams’ was not declared in this scope
  CvSVMParams params;
  ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:304:2: error: ‘params’ was not declared in this scope
  params.svm_type = CvSVM::C_SVC;
  ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:304:20: error: ‘CvSVM’ has not been declared
  params.svm_type = CvSVM::C_SVC;
                    ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:306:26: error: ‘CvSVM’ has not been declared
     params.kernel_type = CvSVM::LINEAR;
                          ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:309:26: error: ‘CvSVM’ has not been declared
     params.kernel_type = CvSVM::POLY;
                          ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:312:3: error: ‘CvSVM’ was not declared in this scope
   CvSVM svm;
   ^
/home/joq/ros/arm/src/agile_grasp/src/agile_grasp/learning.cpp:313:2: error: ‘svm’ was not declared in this scope
  svm.train(features, labels, cv::Mat(), cv::Mat(), params);
  ^
```
```
Warnings   << ml_classifiers:cmake /home/joq/ros/arm/logs/ml_classifiers/build.cmake.000.log
CMake Warning at /opt/ros/kinetic/share/catkin/cmake/catkin_package.cmake:166 (message):
  catkin_package() DEPENDS on 'Eigen3' but neither 'Eigen3_INCLUDE_DIRS' nor
  'Eigen3_LIBRARIES' is defined.
Call Stack (most recent call first):
  /opt/ros/kinetic/share/catkin/cmake/catkin_package.cmake:102 (_catkin_package)
  CMakeLists.txt:18 (catkin_package)
```
```
Warnings   << ml_classifiers:make /home/joq/ros/arm/logs/ml_classifiers/build.make.000.log
/home/joq/ros/arm/src/ml_classifiers/nodes/classifier_server.cpp: In function ‘bool createHelper(std::__cxx11::string, ml_classifiers::Classifier*&)’:
/home/joq/ros/arm/src/ml_classifiers/nodes/classifier_server.cpp:62:52: warning: ‘T* pluginlib::ClassLoader<T>::createClassInstance(const string&, bool) [with T = ml_classifiers::Classifier; std::__cxx11::string = std::__cxx11::basic_string<char>]’ is deprecated [-Wdeprecated-declarations]
         c = c_loader.createClassInstance(class_type);
                                                    ^
In file included from /opt/ros/kinetic/include/pluginlib/./class_loader.hpp:350:0,
                 from /opt/ros/kinetic/include/pluginlib/class_loader.h:35,
                 from /home/joq/ros/arm/src/ml_classifiers/nodes/classifier_server.cpp:50:
/opt/ros/kinetic/include/pluginlib/././class_loader_imp.hpp:129:5: note: declared here
 T * ClassLoader<T>::createClassInstance(const std::string & lookup_name, bool auto_load)
     ^
```
The various Eigen problems are likely due to [some combination of Xenial and Kinetic](http://wiki.ros.org/jade/Migration#Eigen_CMake_Module_in_cmake_modules).

**Comment by jack-oquin** [comment]
There are still a couple of missing dependencies, but they don't look too serious:
```
$ rosdep install --from-paths src --ignore-src --rosdistro=kinetic
ERROR: the following packages/stacks could not have their rosdep keys resolved
to system dependencies:
kinova_arm_moveit_demo: Cannot locate rosdep definition for [pr2_moveit_plugins]
moveit_utils: Cannot locate rosdep definition for [moveit_ros_benchmarks_gui]
```
There is a ``ros-kinetic-moveit-ros-benchmarks`` package, but not with the ``gui`` suffix.

The moveit_utils package is actually contained in segbot_arm.

**Comment by jack-oquin** [comment]
The build dependency on ``ml_classifiers`` is declared in ``segbot_arm_perception``. It does not appear to be used there.

**Comment by jack-oquin** [comment]
Updated .rosinstall, chained to a [Kinetic BWI workspace](https://github.com/utexas-bwi/bwi/blob/master/README.md):
```
- git:
    local-name: agile_grasp
    uri: https://github.com/jsinapov/agile_grasp.git
- git:
    local-name: kinova_ros
    uri: https://github.com/utexas-bwi/kinova-ros.git
- git:
    local-name: segbot_arm
    uri: https://github.com/utexas-bwi/segbot_arm.git
    version: kinetic
```

**Comment by jack-oquin** [comment]
It seems that ``agile_grasp`` has not yet been ported to Kinetic.

**Comment by jsinapov** [comment]
As far as I know, we don't have anything requiring ml_classifiers, I must
have been experimenting with it at one point but that's it. We could always
bug Scott Niekum to update it if we end up needing it :)

Jivko

On Dec 29, 2017 4:31 PM, "jack-oquin" <notifications@github.com> wrote:

The dependency on ml_classifiers is declared as a build depend in
segbot_arm_perception. It does not appear to be used there.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/utexas-bwi/segbot_arm/issues/50#issuecomment-354502736>,
or mute the thread

**Comment by jack-oquin** [comment]
I removed it from the kinetic branch, with no apparent effect on the build.

We do need agile_grasp, which has not been ported to Kinetic and which does not compile there. There is an open issue (atenpas/agile_grasp#5) in the original repo for Kinetic support.

@jsinapov: Your fork has several commits not present in the atenpas repo. How do you suggest we proceed? 

Does it make sense to consolidate our modifications in a utexas-bwi copy? I think we'll need a kinetic branch there for doing the port. Should we submit a pull request to atenpas for that change? For the rest, too?

**Comment by justinhart** [comment]
Another possibility is that we've got GPD pretty close to operational.
We've got a working demo and just need to integrate it with other software
that uses the arm. We could take this as an opportunity to finish that and
get everything going in 1 shot.

**Comment by jack-oquin** [comment]
Is this what you're using? https://github.com/atenpas/gpd

**Comment by jack-oquin** [comment]
segbot_arm has a *lot* of references to agile_grasp. It may be easier to port that than to change all those uses.

**Comment by justinhart** [comment]
That's sensible.

**Comment by jack-oquin** [comment]
I took another run at agile_grasp today. Its ``kinetic`` branch in our fork now compiles and runs the opencv_test, though with an ugly eigen3 deprecation warning in the build.

**Comment by jack-oquin** [comment]
There's a Kinetic version of agile_grasp in the utexas-bwi fork. It builds, but with lots of Eigen warnings.

**Comment by jack-oquin** [comment]
On Kinetic, if a package uses cv_bridge then it must use OpenCV3. With some small CMake hacks, it can still use OpenCV2, but without cv_bridge. See: http://wiki.ros.org/opencv3#Kinetic_and_above

The segbot_arm_perception package uses OpenCV2 and cv_bridge. So, it will need to be ported to OpenCV3, I guess. :frowning_face:

**Comment by jack-oquin** [comment]
For segbot_arm_perception it was just a different include path, see: fd09dcf

**Comment by jack-oquin** [comment]
@nickswalker: I'm through with this for now, and noticed you were working on this stuff, too (thanks!). You might need to pull my latest commit.

**Comment by jack-oquin** [comment]
@nickswalker: what are you doing with kinova_ros? I can compile it, if I put a ``CATKIN_IGNORE`` in kinova_moveit.

**Comment by jack-oquin** [comment]
With the commit above, the new kinetic branch of our kinova-ros fork compiles.

**Comment by jack-oquin** [comment]
We should probably document a procedure that first creates a workspace using the [BWI instructions](https://github.com/utexas-bwi/bwi/blob/master/README.md) and rosinstall files, then adds the (presently three) additional repositories needed for segbot_arm.

That way we won't need to maintain two rosinstall lists for the basic BWI packages.

**Comment by jack-oquin** [comment]
Or, alternatively, we could include segbot_arm in the main BWI repository hierarchy. That probably would need a bit of work to do.

**Comment by nickswalker** [comment]
I'm for moving that process to documentation. I started making that rosinstall because some of the tweaks I've made to the launch files require a different branch of bwi to be checked out. I'm going to have a go at integrating the changes right now to resolve that.

I just made a speculative run at tearing down segbot_arm_launch and integrating it into bwi_launch, segbot_description and segbot_bringup and it went well. I think this will ease things when the arm eventually moves to a v4/v5, and it will avoid duplicating work in dealing with some of the Kinetic updates required (launch files run as is, but with deprecation notices).
